services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/events.py

```diff
@@ -503,114 +506,133 @@ async def action(cls, app: FastAPI, scheduler_data: SchedulerData) -> None:
-                ),
-                scheduler_data.service_name,
-                f"{e}",
-            )
-        app_settings: AppSettings = app.state.settings
-        dynamic_sidecar_settings: DynamicSidecarSettings = (
-            app_settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR
-        )
-
-        # only try to save the status if :
-        # - the dynamic-sidecar has finished booting correctly
-        # - it is requested to save the state
-        if (
-            scheduler_data.dynamic_sidecar.service_removal_state.can_save
-            and await are_all_services_present(
-                node_uuid=scheduler_data.node_uuid,
-                dynamic_sidecar_settings=app.state.settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR,
+        # A node can end up with all the services from a single study.
+        # When the study is closed, all the services will try to save their
+        # data. This causes a lot of disk and network stress.
+        # Some nodes collapse under load or behave unexpectedly.
+        lock_manager = RedisLockManager.instance(app)
+        assert scheduler_data.dynamic_sidecar_node_id  # nosec
+        lock: Optional[Lock] = await lock_manager.acquire_lock(
+            scheduler_data.dynamic_sidecar_node_id
+        )
+        if lock is None:
+            node_slots = await lock_manager.get_node_slots(
+                scheduler_data.dynamic_sidecar_node_id
             )
-        ):
-            dynamic_sidecar_client = get_dynamic_sidecar_client(app)
-            dynamic_sidecar_endpoint = scheduler_data.dynamic_sidecar.endpoint
+            logger.debug("All %s are currently being used. Skipping", f"{node_slots=}")
+            # Next observation cycle, the service will try again to save the state
+            # if it is able to acquire a lock
+            return
 
-            logger.info(
-                "Calling into dynamic-sidecar to save state and pushing data to nodeports"
-            )
+        async with auto_release(lock_manager, lock):
             try:
-                tasks = [
-                    dynamic_sidecar_client.service_push_output_ports(
-                        dynamic_sidecar_endpoint,
-                    )
-                ]
-                # When enabled no longer uploads state via nodeports
-                # S3 is used to store state paths
-                if not app_settings.DIRECTOR_V2_DEV_FEATURES_ENABLED:
-                    tasks.append(
-                        dynamic_sidecar_client.service_save_state(
-                            dynamic_sidecar_endpoint,
-                        )
-                    )
-
-                try:
-                    await logged_gather(*tasks)
-                except NodeportsDidNotFindNodeError as err:
-                    logger.warning("%s", f"{err}")
-
-                logger.info("Ports data pushed by dynamic-sidecar")
+                await dynamic_sidecar_client.begin_service_destruction(
```

Comment:
is this a long running operation?
I just realised that the lock is owned by the director-v2?? why is that?
Why not the dynamic-sidecar directly? especially since it knows on which hostname it lies, the locking would be much more simpler right?

Reply:
> is this a long running operation?

Yes, there are two very long running operations and several shorter ones but still lengthy.

> I just realised that the lock is owned by the director-v2?? why is that?

This is the correct place to lock since:
- the entire block needs to sit under a lock.
- the dynamic scheduler knows best when to save the state. I would not push this responsibility on the sidecar. I would not leave it to each individual sidecar. The current design is as less responsibility as possible to the dynamic-sidecar.
- To correctly check and acquire and release the lock you need to use the same instance.

> Why not the dynamic-sidecar directly? especially since it knows on which hostname it lies, the locking would be much more simpler right?

- locking here is more complex to deal with, since the lock needs to be used across different APIs
- you always need to use the initial distance of the lock to acquire/release it (it must be kept in the app's state)

Reply:
I still think that locking in the dy-sidecar would be way simpler. and also would prevent having so much logic that the director-v2 needs to have.
Just look at real life, the director never knows what the employee is doing exactly. the employee knows best when to lock his drawers.
WEll anyway, this should be moved to the new design when and if it comes.